### PR TITLE
fix/redirection if not logged in store the lastlocation and redirect …

### PR DESCRIPTION
**Pull request by Abhay5855**
### What is the change?
If the user is not logged in and tries to access the intro/:id page, we are redirecting it to the Login page and after successful login, it is redirecting it back to the previous page i.e. intro/:id page instead of the home page.

- In the intro.js we checked if the user is not logged in showing an alert message to redirect him back to the Login page.
- When he tries to access the intro/:id page without login, we are storing the location in session storage so as to access the previous location when logged in.
- On the home page if a user is logged in and if the previous location is present I am redirecting it back to the location which I have stored in the session storage and also removing it as well from storage.

### Is Development Tested?

- [x] Yes
- [ ] No

### Before / After Change Screenshots

Before :

https://github.com/Real-Dev-Squad/website-www/assets/71277980/dae1630b-7df7-489a-a060-01c72aafef3f

After :

https://github.com/Real-Dev-Squad/website-www/assets/71277980/6a6ca3ba-7c6e-4d12-a089-58a510e044fd

### CODE REVIEW

File: js/user.js [hunk 1]
```diff
@@ -26,21 +26,30 @@ const setUserGreeting = (username, firstName, userProfilePicture) => {
...
-    .then((res) => {
-      if (res.error) {
-        throw new Error(res.error);
-      }
-      if (res.incompleteUserDetails) {
-        return window.location.replace('https://my.realdevsquad.com/signup');
-      }
-      setUserGreeting(res.username, res.first_name, res.picture?.url); // BAD
-    });
+  });
```
Comment (Pratiyushkumar) [inline js/user.js:36]: use constant for api url

Comment (Pratiyushkumar) [reply js/user.js:36]: use makeApiCall  function already written in  util.js

**Review by Pratiyushkumar — CHANGES_REQUESTED** [review]

File: index.html [hunk 2]
```diff
@@ -652,5 +652,6 @@ <h3 class="info-note">
     <script src="js/featureFlag.js"></script>
     <script type="module" src="js/map.js"></script>
     <script type="module" src="js/fetchUserData.js" defer></script>
+    <script src="js/utils.js"></script>
   </body>
```
Comment (Pratiyushkumar) [inline index.html:656]: why not put this script above all the files, try to prioritize the scripts

**Review by Pratiyushkumar — APPROVED** [review]
LGTM

**Review by Pratiyushkumar — COMMENTED** [review]
take one more approvals from any member as well

File: js/intro.js [hunk 3]
```diff
@@ -299,7 +299,8 @@ async function showSavedDetails() {
     generatenotAuthorizedPage();
     setTimeout(function () {
       alert('You are not logged in! Redirecting you to login.');
-      window.location.href = HOME_URL;
+      window.location.href = GITHUB_OAUTH;
+      sessionStorage.setItem('lastLocation', window.location.href);
```
Comment (shubhamsigdar1) [inline js/intro.js:303]: Nit: lastLocationUrl

File: js/user.js [hunk 4]
```diff
@@ -26,21 +28,29 @@ const setUserGreeting = (username, firstName, userProfilePicture) => {
...
+    }
+
+    if (result.error) {
+      throw new Error(result.error);
+    }
+
+    if (result.incompleteUserDetails) {
+      return window.location.replace('https://my.realdevsquad.com/signup');
+    }
+    setUserGreeting(result.username, result.first_name, result.picture?.url); // BAD
```
Comment (shubhamsigdar1) [inline js/user.js:50]: why you are repeating setUserGreeting fn two times

**Review by Pratiyushkumar — DISMISSED** [review]
LGTM

**Comment by Abhay5855** [comment]
@shubhamsigdar1 I have done the changes please review once

**Review by shubhamsigdar1 — DISMISSED** [review]

**Review by shubhamsigdar1 — APPROVED** [review]

**Review by sahsisunny — APPROVED** [review]
LGTM

**Review by Pratiyushkumar — DISMISSED** [review]
LGTM

**Review by shubhamsigdar1 — APPROVED** [review]

**Comment by prakashchoudhary07** [comment]
Can we directly pass the redirect URL in AUTH URL it self for redirection

File: js/user.js [hunk 5]
```diff
@@ -26,21 +28,27 @@ const setUserGreeting = (username, firstName, userProfilePicture) => {
...
+    }
+
+    if (result.error) {
+      throw new Error(result.error);
+    }
+
+    if (result.incompleteUserDetails) {
+      return window.location.replace(`${SIGNUP_URL}`);
+    }
+    setUserGreeting(result.username, result.first_name, result.picture?.url); // BAD
```
Comment (harshith-venkatesh) [inline js/user.js:48]: What do you mean by BAD

**Review by harshith-venkatesh — APPROVED** [review]

**Review by sahsisunny — APPROVED** [review]

**Comment by Abhay5855** [comment]
> Can we directly pass the redirect URL in AUTH URL itself for redirection

Hey @prakashchoudhary07 , This might need to add a additional function which does the same thing again and check again if the user is logged in again, which is already done in the user function.

**Pull request merged**